README.md
# CSerial

Cross-platform serial port implementation.

[Homepage](http://programming.rm5248.com/projects/c-serial/)

Based largely off of [JavaSerial](https://github.com/rm5248/JavaSerial),
CSerial aims to bring cross-platform serial programming to a computer near you.

The library is structured along the lines of the Apache Portable Runtime(APR).
It aims to provide as much low-level support as is required to create a working
serial port implementation.

CSerial is designed to be fully POSIX-compliant, however if it is not please
open an issue on GitHub.  See the 'Bugs' section for more details on this.

## Continuous Integration
[![Build Status](https://jenkins.rm5248.com/buildStatus/icon?job=CSerial-buildall
)](https://jenkins.rm5248.com/job/CSerial-buildall)

Find builds on [Jenkins!](https://jenkins.rm5248.com/view/CSerial/)

## Install(Debian-based systems)

There is an APT server setup that provides CSerial that you can find
[here](http://apt.rm5248.com/).  Note that because the 'nightly' repo 
contains packages for both amd64 and armhf, you may need to specify
the exact version to install, as otherwise it may attempt to install 
one from an incorrect arch.

## Build / Install from source

Building is done using CMake.  On Linux, this should be done like the 
following:

```bash
~$ mkdir CSerial_build
~$ cd CSerial_build
~$ cmake /path/to/CSerial_code
~$ make
~$ make install
```

On Windows, I recommend using CMake to create a new Visual Studio project and 
building the code that way.  I have not tested cygwin nor MinGW makefiles as 
generated by CMake.

## Example

For information on how to use the library, check out the [example program](
https://github.com/rm5248/CSerial/blob/master/examples/example_full.c)
under the 'examples' directory.

The simplified flow of how to use the library is below:

1. Allocate a new `c_serial_port_t` struct.  Note that this is an opaque type,
and can only be passed as a pointer.  Use the `c_serial_new` function.
2. Set the name of the serial port.  This is the actual serial port that will 
be opened.  It will be something like `/dev/ttyS0` on Linux systems, and 
`COM1` on Windows systems.  Use the `c_serial_set_port_name` function.
3. Set the serial port parameters.  By default, the port will be opened at 
`9600-8-N-1`.  Any other settings can be set with the appropriate setters.
  * To set the baud rate, use `c_serial_set_baud_rate`
  * To set the number of data bits, use `c_serial_set_data_bits`
  * To set the number of stop bits, use `c_cserial_set_stop_bits`
  * To set the parity, use `c_serial_set_parity`
  * To set the flow control, use `c_serial_set_flow_control`
4. (optional) Set the serial line change flags.  This determines what changes 
on a serial line will cause a read to return.  By default, no flags are set.
Use the `c_serial_set_serial_line_change_flags` function to set the flags.
5. Open the port.  Use `c_serial_open`
6. Read and write data.  Use `c_serial_read_data` and `c_serial_write_data`, 
respectively.

## Logging

The library has some rudimentary logging.  This may be set as either a global 
log function, or as a log function per-port.  There exists a simple 
implementation of a logger which will print out every message to stderr.  This 
is intended for use with a secondary logging framework, so that logs may all 
be used by one logging facility.

The relevant functions for logging are `c_serial_set_log_function` for logging 
a single serial port, and `c_serial_set_global_log_function` for logging 
all messages.  Note that some messages can only go to the global log function, 
as there is not always a serial port associated with some of the operations.

## Other Important Functions

Note that all functions can be see in the Doxygen documentation with 
the proper documentation.  

The following presents a brief list of other functions which may be 
important:

* `c_serial_get_serial_ports_list` - Go through the system and find all of 
the valid serial ports.
* `c_serial_get_last_native_errnum` - Return the last native error number.  
This corresponds to `errno` on POSIX and `GetLastError` on Windows.
* `c_serial_[set|get]_user_data` - Set / get user data(`void*`) associated 
with the port.
* `c_serial_get_available` - Get the number of bytes to read.
* `c_serial_get_native_handle` - Get the native handle(int on POSIX, HANDLE on 
Windows) for raw I/O access
* `c_serial_get_poll_handle` - Get the handle used to poll for changes on a 
port.  Useful for use with a `poll()`-like function.  On POSIX systems, this
returns the same as `c_serial_get_native_handle`.  On Windows, this returns
a `HANDLE` corresponding to an event created with `CreateEvent` on the 
`OVERLAPPED` struct associated with the serial port.

### RTS Handling

The library also supports toggling of the RTS line when transmitting data 
over the serial port.  This is generally used for RS-485 transmitting.  This 
can be set to several different settings(see the enum CSerial_RTS_Handling 
for more details).  If set to HARDWARE, this will attempt to use the built-in 
kernel level functions for controlling the RTS line(using `struct serial_rs485` 
on Linux, or setting `fRtsControl` on Windows).  Setting it to SOFTWARE will 
cause the library to manually toggle the RTS line when sending data.

## Bugs

Questions?  Bugs?  Open an [issue](https://github.com/rm5248/CSerial/issues)
 in GitHub and I will take a look at it.

## License
Licensed under Apache 2.0


debian/changelog
cserial (0.3) unstable; urgency=medium

  * Bumped version after release

 -- robert <robert@debian>  Sat, 29 Jan 2022 12:28:54 -0500

cserial (0.2) unstable; urgency=medium

  * Bumped version for release

 -- robert <robert@robert-macbook>  Sun, 04 Dec 2016 16:37:53 -0500

cserial (0.1) unstable; urgency=medium

  * Initial release of CSerial
  * Cross-platform abilities should be working

 -- robert <robert@debian>  Sun, 21 Aug 2016 14:39:30 -0400

cserial (0) unstable; urgency=low

  * Initial Release.

 -- robert <robert.middleton@rm5248.com>  Fri, 05 Aug 2016 21:44:19 -0400


simplelogger.h
/*
 * Simple logging framework.
 * 
 * License: Unlicense(http://unlicense.org)
 *
 * Find the latest version at https://github.com/rm5248/simplelogger
 *
 * To use: Copy this file into your source directory and define storage for
 * simplelogger_global_log_function in a source file.  Then set your 
 * simplelogger_global_log_function to be a function pointer to a function
 * that you control.
 */
#ifndef SIMPLELOGGER_H
#define SIMPLELOGGER_H

#include "simplelogger_defs.h"

#ifndef SIMPLELOGGER_FUNCTION
  #ifdef _MSC_VER
    #define SIMPLELOGGER_FUNCTION __FUNCSIG__
  #else
    #define SIMPLELOGGER_FUNCTION __func__
  #endif
#endif

/**
 * Global log pointer
 */
#ifndef SIMPLELOGGER_LOG_FUNCTION_NAME
#define SIMPLELOGGER_LOG_FUNCTION_NAME simplelogger_global_log_function
#endif
//extern simplelogger_log_function SIMPLELOGGER_LOG_FUNCTION_NAME;

#define SIMPLELOGGER_LOG_CSTR( logger, message, level ) do{\
    if( !SIMPLELOGGER_LOG_FUNCTION_NAME ) break;\
    struct SL_LogLocation location;\
    location.line_number = __LINE__;\
    location.file = __FILE__;\
    location.function = SIMPLELOGGER_FUNCTION;\
    SIMPLELOGGER_LOG_FUNCTION_NAME( logger, &location, level, message );\
    } while(0)

#define SIMPLELOGGER_TRACE_CSTR( logger, message ) do{\
    SIMPLELOGGER_LOG_CSTR( logger, message, SL_TRACE);\
    } while(0)
#define SIMPLELOGGER_DEBUG_CSTR( logger, message ) do{\
    SIMPLELOGGER_LOG_CSTR( logger, message, SL_DEBUG);\
    } while(0)
#define SIMPLELOGGER_INFO_CSTR( logger, message ) do{\
    SIMPLELOGGER_LOG_CSTR( logger, message, SL_INFO);\
    } while(0)
#define SIMPLELOGGER_WARN_CSTR( logger, message ) do{\
    SIMPLELOGGER_LOG_CSTR( logger, message, SL_WARN);\
    } while(0)
#define SIMPLELOGGER_ERROR_CSTR( logger, message ) do{\
    SIMPLELOGGER_LOG_CSTR( logger, message, SL_ERROR);\
    } while(0)
#define SIMPLELOGGER_FATAL_CSTR( logger, message ) do{\
    SIMPLELOGGER_LOG_CSTR( logger, message, SL_FATAL);\
    } while(0)

#ifdef __cplusplus
#include <string>
#include <sstream>

#define SIMPLELOGGER_TRACE_STDSTR( logger, message ) do{\
    std::stringstream stream;\
    stream << message;\
    SIMPLELOGGER_LOG_CSTR( logger, stream.str().c_str(), SL_TRACE);\
    } while(0)
#define SIMPLELOGGER_DEBUG_STDSTR( logger, message ) do{\
    std::stringstream stream;\
    stream << message;\
    SIMPLELOGGER_LOG_CSTR( logger, stream.str().c_str(), SL_DEBUG);\
    } while(0)
#define SIMPLELOGGER_INFO_STDSTR( logger, message ) do{\
    std::stringstream stream;\
    stream << message;\
    SIMPLELOGGER_LOG_CSTR( logger, stream.str().c_str(), SL_INFO);\
    } while(0)
#define SIMPLELOGGER_WARN_STDSTR( logger, message ) do{\
    std::stringstream stream;\
    stream << message;\
    SIMPLELOGGER_LOG_CSTR( logger, stream.str().c_str(), SL_WARN);\
    } while(0)
#define SIMPLELOGGER_ERROR_STDSTR( logger, message ) do{\
    std::stringstream stream;\
    stream << message;\
    SIMPLELOGGER_LOG_CSTR( logger, stream.str().c_str(), SL_ERROR);\
    } while(0)
#define SIMPLELOGGER_FATAL_STDSTR( logger, message ) do{\
    std::stringstream stream;\
    stream << message;\
    SIMPLELOGGER_LOG_CSTR( logger, stream.str().c_str(), SL_FATAL);\
    } while(0)

#ifdef SIMPLELOGGER_ENABLE_AUTO_MACROS
  #define SIMPLELOGGER_TRACE( logger, message )\
    SIMPLELOGGER_TRACE_STDSTR(logger, message )
  #define SIMPLELOGGER_DEBUG( logger, message )\
    SIMPLELOGGER_DEBUG_STDSTR(logger, message )
  #define SIMPLELOGGER_INFO( logger, message )\
    SIMPLELOGGER_INFO_STDSTR(logger, message )
  #define SIMPLELOGGER_WARN( logger, message )\
    SIMPLELOGGER_WARN_STDSTR(logger, message )
  #define SIMPLELOGGER_ERROR( logger, message )\
    SIMPLELOGGER_ERROR_STDSTR(logger, message )
  #define SIMPLELOGGER_FATAL( logger, message )\
    SIMPLELOGGER_FATAL_STDSTR(logger, message )
#endif /* SIMPLELOGGER_ENABLE_AUTO_MACROS */

#ifdef SIMPLELOGGER_ENABLE_SMALL_MACROS
  #define LOG_TRACE( logger, message )\
    SIMPLELOGGER_TRACE_STDSTR( logger, message )
  #define LOG_DEBUG( logger, message )\
    SIMPLELOGGER_DEBUG_STDSTR( logger, message )
  #define LOG_INFO( logger, message )\
    SIMPLELOGGER_INFO_STDSTR( logger, message )
  #define LOG_WARN( logger, message )\
    SIMPLELOGGER_WARN_STDSTR( logger, message )
  #define LOG_ERROR( logger, message )\
    SIMPLELOGGER_ERROR_STDSTR( logger, message )
  #define LOG_FATAL( logger, message )\
    SIMPLELOGGER_FATAL_STDSTR( logger, message )
#endif /* SIMPLELOGGER_ENABLE_SMALL_MACROS */

#else
/* C macros */

#ifdef SIMPLELOGGER_ENABLE_AUTO_MACROS
  #define SIMPLELOGGER_TRACE( logger, message )\
    SIMPLELOGGER_TRACE_CSTR(logger, message )
  #define SIMPLELOGGER_DEBUG( logger, message )\
    SIMPLELOGGER_DEBUG_CSTR(logger, message )
  #define SIMPLELOGGER_INFO( logger, message )\
    SIMPLELOGGER_INFO_CSTR(logger, message )
  #define SIMPLELOGGER_WARN( logger, message )\
    SIMPLELOGGER_WARN_CSTR(logger, message )
  #define SIMPLELOGGER_ERROR( logger, message )\
    SIMPLELOGGER_ERROR_CSTR(logger, message )
  #define SIMPLELOGGER_FATAL( logger, message )\
    SIMPLELOGGER_FATAL_CSTR(logger, message )
#endif /* SIMPLELOGGER_ENABLE_AUTO_MACROS */

#ifdef SIMPLELOGGER_ENABLE_SMALL_MACROS
  #define LOG_TRACE( logger, message )\
    SIMPLELOGGER_TRACE_CSTR( logger, message )
  #define LOG_DEBUG( logger, message )\
    SIMPLELOGGER_DEBUG_CSTR( logger, message )
  #define LOG_INFO( logger, message )\
    SIMPLELOGGER_INFO_CSTR( logger, message )
  #define LOG_WARN( logger, message )\
    SIMPLELOGGER_WARN_CSTR( logger, message )
  #define LOG_ERROR( logger, message )\
    SIMPLELOGGER_ERROR_CSTR( logger, message )
  #define LOG_FATAL( logger, message )\
    SIMPLELOGGER_FATAL_CSTR( logger, message )
#endif /* SIMPLELOGGER_ENABLE_SMALL_MACROS */

#endif /* __cplusplus */


#endif /* SIMPLELOGGER_H */


examples/example_full.c
/*
   Copyright 2016 rm5248

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
*/

#include <stdint.h>
#include <stdio.h>
#include <cserial/c_serial.h>

/**
 */

int main( int argc, char** argv ){
    c_serial_port_t* m_port;
    c_serial_control_lines_t m_lines;
    int status;
    int bytes_read;
    uint8_t data[ 255 ];
    int data_length;
    int x;

    /*
     * Use the first argument as the port to open
     */
    if( argc != 2 ){
        fprintf( stderr, "ERROR: First argument must be serial port\n" );
		/*
		 * Since no port was provided, print the available ports
		 */
		const char** port_list = c_serial_get_serial_ports_list();
		x = 0;
		printf("Available ports:\n");
		while( port_list[ x ] != NULL ){
			printf( "%s\n", port_list[ x ] );
			x++;
		}
		c_serial_free_serial_ports_list( port_list );
        return 1;
    }
 
    /*
     * Set the global log function.
     * We will use a simple log function that prints to stderr
     */
    c_serial_set_global_log_function( c_serial_stderr_log_function );

    /*
     * Allocate the serial port struct.
     * This defaults to 9600-8-N-1
     */
    if( c_serial_new( &m_port, NULL ) < 0 ){
        fprintf( stderr, "ERROR: Unable to create new serial port\n" );
        return 1;
    }

    /*
     * The port name is the device to open(/dev/ttyS0 on Linux,
     * COM1 on Windows)
     */
    if( c_serial_set_port_name( m_port, argv[ 1 ] ) < 0 ){
        fprintf( stderr, "ERROR: can't set port name\n" );
    }

    /*
     * Set various serial port settings.  These are the default.
     */
    c_serial_set_baud_rate( m_port, CSERIAL_BAUD_9600 );
    c_serial_set_data_bits( m_port, CSERIAL_BITS_8 );
    c_serial_set_stop_bits( m_port, CSERIAL_STOP_BITS_1 );
    c_serial_set_parity( m_port, CSERIAL_PARITY_NONE );
    c_serial_set_flow_control( m_port, CSERIAL_FLOW_NONE );

    printf( "Baud rate is %d\n", c_serial_get_baud_rate( m_port ) );

    /*
     * We want to get all line flags when they change
     */
    c_serial_set_serial_line_change_flags( m_port, CSERIAL_LINE_FLAG_ALL );

    status = c_serial_open( m_port );
    if( status < 0 ){
        fprintf( stderr, "ERROR: Can't open serial port\n" );
        return 1;
    }

    /*
     * Listen for anything that comes across, and echo it back.
     */
    do{
        data_length = 255;
        status = c_serial_read_data( m_port, data, &data_length, &m_lines );
        if( status < 0 ){
            break;
        }


        printf( "Got %d bytes of data\n", data_length );
        for( x = 0; x < data_length; x++ ){
            printf( "    0x%02X (ASCII: %c)\n", data[ x ], data[ x ] );
        }
        printf( "Serial line state: CD: %d CTS: %d DSR: %d DTR: %d RTS: %d RI: %d\n",
            m_lines.cd, 
            m_lines.cts,
            m_lines.dsr,
            m_lines.dtr,
            m_lines.rts,
            m_lines.ri );

        status = c_serial_write_data( m_port, data, &data_length );
        if( status < 0 ){
            break;
        }
    }while( 1 );
}


c_serial.h
/*
   Copyright 2016 rm5248

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
*/
#ifndef C_SERIAL_H
#define C_SERIAL_H

#include "simplelogger_defs.h"

/**
 * \addtogroup CSerial
 * @{
 */

#ifdef __cplusplus
extern "C" {
#endif /* __cplusplus */

#ifdef _WIN32
 
  #ifdef cserial_EXPORTS
    #define CSERIAL_EXPORT __declspec( dllexport )
  #else
    #define CSERIAL_EXPORT __declspec( dllimport )
  #endif /* CSERIAL_LIB */

	/* Static libraries don't use dllexport/dllimport */
#ifdef CSERIAL_STATIC
#undef CSERIAL_EXPORT
#define CSERIAL_EXPORT
#endif

#include <windows.h>

typedef HANDLE c_serial_handle_t;
typedef DWORD c_serial_errnum_t;

#else
    /* Linux/POSIX zone */
    #define CSERIAL_EXPORT

typedef int c_serial_handle_t;
typedef int c_serial_errnum_t;

#endif /* _WIN32 */

/*
 * Error code definitions
 */
/** Everything OK */
#define CSERIAL_OK 0
/** Generic error */
#define CSERIAL_ERROR_GENERIC -1
/** The port we tried to open is not actually a serial port */
#define CSERIAL_ERROR_NOT_A_SERIAL_PORT -2
/** The port that we tried to open does not exist */
#define CSERIAL_ERROR_NO_SUCH_SERIAL_PORT -3
/** The parameters to cserial_read were incorrect */
#define CSERIAL_ERROR_INCORRECT_READ_PARAMS -4
/** Unable to create new serial port */
#define CSERIAL_ERROR_CANT_CREATE -5
/** No port defined to open */
#define CSERIAL_ERROR_NO_PORT -6
/** Port already open */
#define CSERIAL_ERROR_ALREADY_OPEN -7
/** invalid c_serial_port_t, e.g. it is NULL */
#define CSERIAL_ERROR_INVALID_PORT -8
/** name of the port to open is too long(6 chars on Windows, 255 on POSIX */
#define CSERIAL_ERROR_NAME_TOO_LONG -9
/** Invalid parameters for flow control. This will be returned if the flow
 * control and the RTS handling conflict with each other.
 */
#define CSERIAL_ERROR_INVALID_FLOW -10
/** Returned if the specified RTS control type is not available
 */
#define CSERIAL_RTS_TYPE_NOT_AVAILABLE -11

/**
 * Serial line flags
 */
#define CSERIAL_LINE_FLAG_CTS     ( 0x01 << 0 )
#define CSERIAL_LINE_FLAG_DSR     ( 0x01 << 1 )
#define CSERIAL_LINE_FLAG_DTR     ( 0x01 << 2 )
#define CSERIAL_LINE_FLAG_RTS     ( 0x01 << 3 )
#define CSERIAL_LINE_FLAG_RI      ( 0x01 << 4 )
#define CSERIAL_LINE_FLAG_CD      ( 0x01 << 5 )
#define CSERIAL_LINE_FLAG_NONE    0x00
#define CSERIAL_LINE_FLAG_ALL     ( CSERIAL_LINE_FLAG_CTS | \
                                    CSERIAL_LINE_FLAG_DSR | \
                                    CSERIAL_LINE_FLAG_DTR | \
                                    CSERIAL_LINE_FLAG_RTS | \
                                    CSERIAL_LINE_FLAG_RI  | \
                                    CSERIAL_LINE_FLAG_CD )

/**
 * Struct representing the control line state.
 */
struct c_serial_control_lines {
    /** CD - Carrier Detect */
    int cd;
    /** CTS - Clear To Send */
    int cts;
    /** DSR - Data Set Ready */
    int dsr;
    /** DTR - Data Terminal Ready */
    int dtr;
    /** RTS - Request To Send */
    int rts;
    /** Ring Indicator(is the phone ringing?) */
    int ri;
};
typedef struct c_serial_control_lines c_serial_control_lines_t;

/**
 * Enum for baud rate
 */
enum CSerial_Baud_Rate {
    /** Not supported by Windows */
    CSERIAL_BAUD_0 = 0,
    /** Not supported by Windows */
    CSERIAL_BAUD_50 = 50,
    /** Not supported by Windows */
    CSERIAL_BAUD_100 = 100,

    CSERIAL_BAUD_110 = 110,

    /** Not supported by Windows */
    CSERIAL_BAUD_134 = 134,
    /** Not supported by Windows */
    CSERIAL_BAUD_150 = 150,
    /** Not supported by Windows */
    CSERIAL_BAUD_200 = 200,
    
    CSERIAL_BAUD_300 = 300,
    CSERIAL_BAUD_600 = 600,
    CSERIAL_BAUD_1200 = 1200,

    /** Not supported by Windows */
    CSERIAL_BAUD_1800 = 1800,
 
    CSERIAL_BAUD_2400 = 2400,
    CSERIAL_BAUD_4800 = 4800,
    CSERIAL_BAUD_9600 = 9600,
    CSERIAL_BAUD_19200 = 19200,
    CSERIAL_BAUD_38400 = 38400,
    CSERIAL_BAUD_57600 = 57600,
    CSERIAL_BAUD_115200 = 115200
}; 

enum CSerial_Data_Bits{
    CSERIAL_BITS_5 = 5,
    CSERIAL_BITS_6,
    CSERIAL_BITS_7,
    CSERIAL_BITS_8
};

enum CSerial_Stop_Bits{
    CSERIAL_STOP_BITS_1 = 1,
    CSERIAL_STOP_BITS_2
};

enum CSerial_Parity{
    CSERIAL_PARITY_NONE = 0,
    CSERIAL_PARITY_ODD,
    CSERIAL_PARITY_EVEN
};

enum CSerial_Flow_Control{
    CSERIAL_FLOW_NONE = 0,
    CSERIAL_FLOW_HARDWARE,
    CSERIAL_FLOW_SOFTWARE
};

enum CSerial_RTS_Handling{
    /** Do not handle the RTS line */
    CSERIAL_RTS_NONE = 0,
    /** Use hardware/driver level RTS handling.   */
    CSERIAL_RTS_HARDWARE,
    /** Use software-level RTS handling.  This will cause any calls 
     * to c_serial_write_data to block
     */
    CSERIAL_RTS_SOFTWARE,
    /** Use the best available RTS handling.
     * CSERIAL_RTS_HARDWARE will be used if it is available
     */
    CSERIAL_RTS_BEST_AVAILABLE
};

/*
 * Opaque type for interfacing with a serial port.
 */
typedef struct c_serial_port c_serial_port_t;

/**
 * Cerate a new C Serial object with default settings.
 * 
 * @param port A pointer to the pointer that will be filled in with the 
 *  new data.
 * @param errnum A pointer to the native error type if extended error
 *  information is required
 * @return CSERIAL_OK or an error code.  If there was an error, the port will
 *  not be valid.
 */
CSERIAL_EXPORT int c_serial_new( c_serial_port_t** port,
                                 c_serial_errnum_t* errnum );

/**
 * Close the serial port(if it is not already closed) and free all resources.
 * The pointer will be invalid after this call
 */
CSERIAL_EXPORT void c_serial_free( c_serial_port_t* port );

/**
 * Close the port associated with this serial port.  The port may be opened
 * again by calling c_serial_open()
 */
CSERIAL_EXPORT void c_serial_close( c_serial_port_t* port );

/**
 * Open the port on the system.
 */
CSERIAL_EXPORT int c_serial_open( c_serial_port_t* port );

/**
 * Open the port on the system, optionally keeping all settings.
 * If keepSettings is set to 0, acts the same as c_serial_open
 * 
 * @param port The port to open
 * @param keepSettings 1 if current serial port settings should be kept,
 *  0 otherwise
 */
CSERIAL_EXPORT int c_serial_open_keep_settings( c_serial_port_t* port,
                                                int keepSettings );

/**
 * Returns 1 if the port is open, 0 otherwise
 */
CSERIAL_EXPORT int c_serial_is_open( c_serial_port_t* port );

/**
 * Set the name of the port to open.
 * The value sent is platform-dependent, e.g. COM1 on Windows and /dev/ttyS1
 * on Linux and Linux-like systems.
 *
 * The port name is copied to internal memory and may be freed after.
 */
CSERIAL_EXPORT int c_serial_set_port_name( c_serial_port_t* port,
                                           const char* port_name );

/**
 * Get the port name that this serial port represents, e.g. COM1 on Windows
 * and /dev/ttyS1 on Linux and Linux-like systems
 */
CSERIAL_EXPORT const char* c_serial_get_port_name( c_serial_port_t* port );

/**
 * Set the baud rate of the serial port
 */
CSERIAL_EXPORT int c_serial_set_baud_rate( c_serial_port_t* port, 
                                           enum CSerial_Baud_Rate baud );

/**
 * Get the baud rate of the serial port.  Note that this function returns 
 * a cached value if the port is not open, otherwise it reads directly
 * from the port and returns the proper value.
 */
CSERIAL_EXPORT enum CSerial_Baud_Rate c_serial_get_baud_rate( 
                                               c_serial_port_t* port );

/**
 * Set the number of data bits.
 */
CSERIAL_EXPORT int c_serial_set_data_bits( c_serial_port_t* port,
                                           enum CSerial_Data_Bits bits );

/**
 * Get the number of data bits
 */
CSERIAL_EXPORT enum CSerial_Data_Bits c_serial_get_data_bits( 
                                                   c_serial_port_t* port );

/**
 * Set the number of stop bits
 */
CSERIAL_EXPORT int c_serial_set_stop_bits( c_serial_port_t* port,
                                           enum CSerial_Stop_Bits bits );

/**
 * Get the number of stop bits
 */
CSERIAL_EXPORT enum CSerial_Stop_Bits c_serial_get_stop_bits(
                                                   c_serial_port_t* port );

/**
 * Set the parity 
 */
CSERIAL_EXPORT int c_serial_set_parity( c_serial_port_t* port,
                                        enum CSerial_Parity parity );

/**
 * Get the parity.
 */
CSERIAL_EXPORT enum CSerial_Parity c_serial_get_parity( 
                                                  c_serial_port_t* port );

/**
 * Set the flow control
 */
CSERIAL_EXPORT int c_serial_set_flow_control( c_serial_port_t* port,
                                             enum CSerial_Flow_Control contol );

/**
 * Get the flow control
 */
CSERIAL_EXPORT enum CSerial_Flow_Control c_serial_get_flow_control(
                                                  c_serial_port_t* port );

/**
 * Write a block of data to the serial port.
 * Acts the same as write() in POSIX systems.
 *
 * @param port The port to write data out to.
 * @param data The data to write out to the port.
 * @param length How long(in bytes) the data is.  Set to the number of bytes written on return.
 * @return status code
 */
CSERIAL_EXPORT int c_serial_write_data( c_serial_port_t* port,
                                        void* data,
                                        int* length );

/**
 * Read data from the serial port.
 * Acts like read() in POSIX systems, except that it will also return the 
 * state of the serial lines.  This function will block for data before it 
 * returns.  If a mask of serial lines to listen to has been set using
 * c_serial_set_serial_line_change_flags, then this function may return
 * without having read any data into the data buffer.
 *
 * If data is NULL and lines is non-NULL, but no change_flags have been set,
 * this function returns immediately without having read any data.
 * 
 * @param port The port to read data from
 * @param data The buffer to write data to.  May be NULL if you are only
 *  interested in listening for serial line changes.
 * @param data_length On entry, how long the data is.  On exit, how many bytes
 *  were read. 
 * @param lines The state of the serial lines on read of data.  If waiting for
 *  the line state to change, cannot be NULL
 * @return CSERIAL_OK on success, error code otherwise
 */
CSERIAL_EXPORT int c_serial_read_data( c_serial_port_t* port,
                                       void* data,
                                       int* data_length,
                                       c_serial_control_lines_t* lines );

/**
 * Get the native handle(int or HANDLE) that is used by this serial port.
 * 
 * Note that on Windows, this will return the actual handle used for 
 * the ReadFile/WriteFile system calls.  To get the HANDLE for the 
 * Event, use c_serial_get_poll_handle()
 */
CSERIAL_EXPORT c_serial_handle_t c_serial_get_native_handle( 
                                                 c_serial_port_t* port );

/**
 * Get the native handle used for a poll()-like function.
 * On POSIX systems, this will return the same as 
 * c_serial_get_native_handle()
 *
 * On Windows, this will return a HANDLE to an created with the 
 * CreateEvent system call, which can then be used to listen for changes
 * using a function such as MsgWaitForMultipleObjectsEx or 
 * WaitForSingleObject.
 */
CSERIAL_EXPORT c_serial_handle_t c_serial_get_poll_handle(
                                                 c_serial_port_t* port );

/**
 * Set the state of the control lines, optionally returning the current 
 * state of the lines after setting them.
 * 
 * @param port The port to set the line state for.
 * @param lines The line state to set.  Note that only DTR and RTS can be set.
 * @param return_state If true, modify lines on return with the current state
 *  of the serial lines.
 */
CSERIAL_EXPORT int c_serial_set_control_line( c_serial_port_t* port,
                                              c_serial_control_lines_t* lines,
                                              int return_state );

/**
 * Get the current state of the serial lines
 */
CSERIAL_EXPORT int c_serial_get_control_lines( c_serial_port_t* port,
                                              c_serial_control_lines_t* lines );

/**
 * Get the number of bytes currently available
 * 
 * @param port The port to get the number of bytes currently available from
 * @param available The number of bytes available
 * @return CSERIAL_OK or an error code
 */
CSERIAL_EXPORT int c_serial_get_available( c_serial_port_t* port,
                                           int* available );

/**
 * Set the serial lines that will cause a return from c_serial_read
 *
 * @param port The port to set the line bitmask for
 * @param flags Bitwise-OR of any CSERIAL_LINE_FLAG_XXX
 */
CSERIAL_EXPORT int c_serial_set_serial_line_change_flags( c_serial_port_t* port,
                                                          int flags );

/**
 * Get the serial line flags that will cause a return from c_serial_read
 */
CSERIAL_EXPORT int c_serial_get_serial_line_change_flags( 
                                                    c_serial_port_t* port );

/**
 * Set any user data that should be associated with this serial port 
 */
CSERIAL_EXPORT void c_serial_set_user_data( c_serial_port_t* port, void* data );

/**
 * Get any user data associated with this serial port
 */
CSERIAL_EXPORT void* c_serial_get_user_data( c_serial_port_t* port );

/**
 * Get the text of a C Serial error number
 */
CSERIAL_EXPORT const char* c_serial_get_error_string( int errnum );

/**
 * Set the global log function that will get all of the 
 */
CSERIAL_EXPORT int c_serial_set_global_log_function( 
                                                 simplelogger_log_function func );

/**
 * A simple implementation of a log function which outputs log information
 * to stderr.
 */
CSERIAL_EXPORT void c_serial_stderr_log_function(const char* logger_name, 
                                                 const struct SL_LogLocation* location,
                                                 const enum SL_LogLevel level,
                                                 const char* log_string );

/**
 * Get the last error number associated with this port.  This corresponds
 * to errno on Linux-like systems, and GetLastError() on Windows.
 */
CSERIAL_EXPORT c_serial_errnum_t c_serial_get_last_native_errnum( 
                                                       c_serial_port_t* port );

/**
 * Get a list of all serial ports that are on the system.
 * This must be freed with c_serial_free_serial_ports_list()
 *
 * @return An array of char* listing all of the serial ports on the system
 *  e.g. /dev/ttyS1, /dev/ttyS2 on Linux; COM1,COM2 on Windows.  This array is 
 *  NULL-terminated.
 */
CSERIAL_EXPORT const char** c_serial_get_serial_ports_list();

/**
 * Free the list of serial ports retrieved with c_serial_get_serial_ports
 */
CSERIAL_EXPORT void c_serial_free_serial_ports_list( const char** port_list );

/**
 * Set how RTS should be controlled.  This is generally used for RS-485 
 * converters.  Note that this logically separate from the flow control, even
 * if they both use RTS.  That is, if c_serial_set_flow_control() is set with 
 * HARDWARE flow control, this will override that setting.
 *
 * Important note regarding setting the RTS control: If the port is not open,
 * this function may return CSERIAL_OK if the setting is valid, and upon
 * opening the port CSERIAL_RTS_TYPE_NOT_AVAILABLE may be returned.
 *
 * @param handling How to handle setting the RTS control.  HARDWARE indicates 
 * that we will attempt to use the OS-level way of setting the RTS line.
 * SOFTWARE indicates that we will set the RTS line on the library side.
 * NONE turns it off.
 */
CSERIAL_EXPORT int c_serial_set_rts_control( c_serial_port_t* port,
                                        enum CSerial_RTS_Handling handling );

/**
 * Get the current handling of the RTS line for RS485 handling
 */
CSERIAL_EXPORT enum CSerial_RTS_Handling c_serial_get_rts_control( 
                                                       c_serial_port_t* port );

/**
 * Flush all data.
 */
CSERIAL_EXPORT int c_serial_flush( c_serial_port_t* port );

#ifdef __cplusplus
} /* end extern "C" */
#endif /* __cplusplus */

/** @} */

#endif /* C_SERIAL_H */
